test/12_dependency_versions/cibuildwheel_test.py

```diff
@@ -0,0 +1,115 @@
+import os
+import re
+import pytest
+import textwrap
+import cibuildwheel.util
+
+import utils
+
+VERSION_REGEX = r'([\w-]+)==([^\s]+)'
+
+
+def get_versions_from_constraint_file(constraint_file):
+    with open(constraint_file, encoding='utf8') as f:
+        constraint_file_text = f.read()
+
+    versions = {}
+
+    for package, version in re.findall(VERSION_REGEX, constraint_file_text):
+        versions[package] = version
+
+    return versions
+
+
+@pytest.mark.parametrize('python_version', ['2.7', '3.x'])
+def test_pinned_versions(python_version):
+    if utils.platform == 'linux':
+        pytest.skip('linux doesn\'t pin individual tool versions, it pins manylinux images instead')
+
+    project_dir = os.path.dirname(__file__)
+
+    build_environment = {}
+
+    if python_version == '2.7':
+        constraint_filename = 'constraints-python27.txt'
+    else:
+        constraint_filename = 'constraints.txt'
+
+    constraint_file = os.path.join(cibuildwheel.util.resources_dir, constraint_filename)
+    constraint_versions = get_versions_from_constraint_file(constraint_file)
+
+    for package in ['pip', 'setuptools', 'wheel', 'virtualenv']:
+        env_name = 'EXPECTED_{}_VERSION'.format(package.upper())
+        build_environment[env_name] = constraint_versions[package]
+
+    cibw_environment_option = ' '.join(
+        ['{}={}'.format(k, v) for k, v in build_environment.items()]
+    )
+
+    # build and test the wheels
+    actual_wheels = utils.cibuildwheel_run(project_dir, add_env={
+        'CIBW_BUILD': '[cp]p27-*' if python_version == '2.7' else '[cp]p3?-*',
+        'CIBW_ENVIRONMENT': cibw_environment_option,
+    })
+
+    # also check that we got the right wheels
+    if python_version == '2.7':
+        expected_wheels = [w for w in utils.expected_wheels('spam', '0.1.0')
+                           if '-cp27' in w or '-pp27' in w]
```

Comment:
`'27' in w` or `'p27' in w`?

similar below?

Reply:
I'm not sure your suggestions are clearer, because they have even less context for which part of the filename we should be looking for. these expected_wheel comprehensions always bamboozle me, to be honest. Would this be better?

```python
expected_wheels = fnmatch.filter(utils.expected_wheels('spam', '0.1.0'), '*-[cp]p27-*')
```

Reply:
yes

Reply:
To me, the `fnmatch.filter` seems actually more complex (and less flexible for future additions) than just the two cases, but it also works.

Reply:
yeah, on reflection I agree with @YannickJadoul on this one. i'll leave as-is.

Reply:
but in this case variable `constraint_file` from previous test is obsolete.